src/query_plan/operation.rs

```diff
@@ -1459,17 +1970,95 @@ impl NormalizedFieldSelection {
                 sibling_typename: None,
             }),
             selection_set: if field_composite_type_result.is_ok() {
-                Some(NormalizedSelectionSet::normalize_and_expand_fragments(
+                Some(NormalizedSelectionSet::from_selection_set(
                     &field.selection_set,
                     fragments,
                     schema,
-                    normalize_fragment_spread_option,
                 )?)
             } else {
                 None
             },
         }))
     }
+
+    pub(crate) fn normalize(
+        &self,
+        _parent_type: &CompositeTypeDefinitionPosition,
+    ) -> NormalizedFieldSelection {
+        // TODO - will be fixed in subsequent PR
+        self.clone()
+    }
+
+    /// Returns a field selection "equivalent" to the one represented by this object, but such that its parent type
+    /// is the one provided as argument.
+    ///
+    /// Obviously, this operation will only succeed if this selection (both the field itself and its subselections)
+    /// make sense from the provided parent type. If this is not the case, this method will throw.
+    pub(crate) fn rebase_on(
+        &self,
+        parent_type: &CompositeTypeDefinitionPosition,
+        named_fragments: &NamedFragments,
+        schema: &ValidFederationSchema,
+        error_handling: &RebaseErrorHandlingOption,
+    ) -> Result<Option<NormalizedSelection>, FederationError> {
+        if &self.field.data().schema == schema
+            && &self.field.data().field_position.parent() == parent_type
+        {
+            return Ok(Some(NormalizedSelection::Field(Arc::new(self.clone()))));
+        }
+
+        if let Some(rebased) = self.field.rebase_on(parent_type, schema, error_handling)? {
+            if let Some(selection_set) = &self.selection_set {
```

Comment:
You can mimick the JS implementation where it uses an early return and save a level of nesting with `let-else` syntax here
```rust
let Some(selection_set) = &self.selection_set else {
    return Ok(None);
};
```

Reply:
[f985b43](https://github.com/apollographql/federation-next/pull/239/commits/f985b430bb094044f6ea121bb4ee1691fd11ee1e)